src/java/org/apache/cassandra/db/memtable/Memtable.java

```diff
@@ -0,0 +1,389 @@
+     * or a static {@code factory(Map<String, String>)} method. In the latter case, the method should avoid creating
+     * multiple instances of the factory for the same parameters, or factories should at least implement hashCode and
+     * equals.
+     */
+    interface Factory
+    {
+        /**
+         * Create a memtable.
+         *
+         * @param commitLogLowerBound A commit log lower bound for the new memtable. This will be equal to the previous
+         *                            memtable's upper bound and defines the span of positions that any flushed sstable
+         *                            will cover.
+         * @param metadaRef Pointer to the up-to-date table metadata.
+         * @param owner Owning objects that will receive flush requests triggered by the memtable (e.g. on expiration).
+         */
+        Memtable create(AtomicReference<CommitLogPosition> commitLogLowerBound, TableMetadataRef metadaRef, Owner owner);
+
+        /**
+         * If the memtable can achieve write durability directly (i.e. using some feature other than the commitlog, e.g.
+         * persistent memory), it can return true here, in which case the commit log will not store mutations in this
+         * table.
+         * Note that doing so will prevent point-in-time restores and changed data capture, thus a durable memtable must
+         * allow the option of turning commit log writing on even if it does not need it.
+         */
+        default boolean writesShouldSkipCommitLog()
+        {
+            return false;
+        }
+
+        /**
+         * This should be true if the memtable can achieve write durability for crash recovery directly (i.e. using some
+         * feature other than the commitlog, e.g. persistent memory).
+         * Setting this flag to true means that the commitlog should not replay mutations for this table on restart,
+         * and that it should not try to preserve segments that contain relevant data.
+         * Unless writesShouldSkipCommitLog() is also true, writes will be recorded in the commit log as they may be
+         * needed for changed data capture or point-in-time restore.
+         */
+        default boolean writesAreDurable()
+        {
+            return false;
+        }
+
+        /**
+         * Normally we can receive streamed sstables directly, but if the memtable can request to receive the data
+         * instead by returning true here.
+         */
+        default boolean streamToMemtable()
+        {
+            return false;
+        }
+
+        /**
+         * When we need to stream data, we usually flush and stream the resulting sstables. If the sstable does not
+         * want to flush, it should return true here and false on shouldSwitch(STREAMING/REPAIR).
+         * In that case, temporary sstables will be created, streamed and then deleted.
+         */
+        default boolean streamFromMemtable()
```

Comment:
What is actually the requirement for being able to stream from memtable?

Is it enough if the data is persisted in the commit log? Or must the memtable itself provide persitency of the writes? 
Or there is no such requirement because temporary sstables will be created, streamed and then deleted?

What happens if memtable returns `true` on `shouldSwitch` but supports streaming from itself? Should we put an assert somewhere that ensures conisntency of the logic?

Reply:
There are no requirements -- any memtable would be able to handle both streaming from and to it. It is not an error to do so, though it is usually less efficient.
However, if the memtable does not want to flush, then it is necessary to stream from/to the memtable.